sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -506,30 +506,55 @@ void replaceOutputs(Map<PValue, ReplacementOutput> originalToReplacement) {
      * <p>Provides an ordered visit of the input values, the primitive transform (or child nodes for
      * composite transforms), then the output values.
      */
```

Comment:
perhaps beefier comment explaining expected traversal behavior? the logic here seems sufficiently complex that it would help explain exactly why we do what we do.